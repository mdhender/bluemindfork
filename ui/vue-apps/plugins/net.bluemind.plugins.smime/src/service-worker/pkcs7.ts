import { pkcs7, pki, asn1, util } from "node-forge";

export async function decrypt(
    data: Blob,
    privateKey: pki.rsa.PrivateKey,
    certificate: pki.Certificate
): Promise<string | undefined> {
    const buffer = await data.arrayBuffer();
    const text = asn1.fromDer(new util.ByteStringBuffer(buffer));
    const envelope = <pkcs7.Captured<pkcs7.PkcsEnvelopedData>>pkcs7.messageFromAsn1(text);
    const recipient = envelope.findRecipient(certificate);
    if (recipient) {
        envelope.decrypt(recipient, privateKey);
        return envelope.content?.toString();
    }
}

export async function crypt() {
    return null;
}

export async function verify() {
    return null;
}

export async function sign() {
    return null;
}

export default { decrypt, crypt, verify, sign };
