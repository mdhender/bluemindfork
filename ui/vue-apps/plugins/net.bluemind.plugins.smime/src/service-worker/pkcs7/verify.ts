import { md, pkcs7, pki, asn1, util } from "node-forge";
import {
    InvalidMessageIntegrityError,
    InvalidPkcs7EnvelopeError,
    InvalidSignatureError,
    UnsupportedAlgorithmError
} from "../../lib/exceptions";

export function getSigningTime(envelope: pkcs7.Captured<pkcs7.PkcsSignedData>) {
    const parent = getParent(envelope.rawCapture.authenticatedAttributesAsn1, pki.oids["signingTime"]);
    if (!parent) {
        throw new InvalidPkcs7EnvelopeError();
    }
    return asn1.utcTimeToDate(<string>(<asn1.Asn1>(<asn1.Asn1>parent.value[1]).value[0]).value);
}

export function checkSignatureValidity(envelope: pkcs7.Captured<pkcs7.PkcsSignedData>, certificate: pki.Certificate) {
    const authAttributesDigest = digestAttributes(envelope.rawCapture.authenticatedAttributesAsn1, certificate);

    const pubKey = <pki.rsa.PublicKey>certificate.publicKey;
    if (!pubKey.verify(authAttributesDigest, envelope.rawCapture.signature, "RSASSA-PKCS1-V1_5")) {
        throw new InvalidSignatureError();
    }
}

export function checkMessageIntegrity(envelope: pkcs7.Captured<pkcs7.PkcsSignedData>, contentToDigest: string) {
    const algo = getHashFunction(asn1.derToOid(envelope.rawCapture.digestAlgorithm));
    const computedDigest = algo.create().update(contentToDigest).digest().bytes();
    const messageDigest = getMessageDigest(envelope.rawCapture.authenticatedAttributesAsn1);
    if (util.bytesToHex(<string>messageDigest) !== util.bytesToHex(computedDigest)) {
        if (contentToDigest.match(/(?<!\r)\n/) === null) {
            throw new InvalidMessageIntegrityError();
        }
        contentToDigest = contentToDigest.replace(/(?<!\r)\n/g, "\r\n");
        const newDigest = algo.create().update(contentToDigest).digest().bytes();
        if (util.bytesToHex(<string>messageDigest) !== util.bytesToHex(newDigest)) {
            throw new InvalidMessageIntegrityError();
        }
    }
}

function digestAttributes(attrs: pki.Attribute, certificate: pki.Certificate): string {
    // per RFC 2315, attributes are to be digested using a SET container
    // not the above [0] IMPLICIT container
    const attrsAsn1 = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SET, true, []);
    attrsAsn1.value = attrs.value;

    // DER-serialize and digest SET OF attributes only
    const bytes = asn1.toDer(attrsAsn1).getBytes();

    const algo = getDigestAttrAlgo(certificate.siginfo.algorithmOid);
    return algo.update(bytes).digest().getBytes();
}

function getDigestAttrAlgo(algorithmOid: string) {
    switch (algorithmOid) {
        case pki.oids.md5WithRSAEncryption:
            return md.md5.create();
        case pki.oids.sha1WithRSAEncryption:
            return md.sha1.create();
        case pki.oids.sha256WithRSAEncryption:
            return md.sha256.create();
        case pki.oids.sha384WithRSAEncryption:
            return md.sha384.create();
        case pki.oids.sha512WithRSAEncryption:
            return md.sha512.create();
        default:
            throw new UnsupportedAlgorithmError(pki.oids[algorithmOid]);
    }
}

function getHashFunction(algorithmOid: string) {
    switch (algorithmOid) {
        case pki.oids.md5:
            return md.md5;
        case pki.oids.sha1:
            return md.sha1;
        case pki.oids.sha256:
            return md.sha256;
        case pki.oids.sha384:
            return md.sha384;
        case pki.oids.sha512:
            return md.sha512;
        default:
            throw new UnsupportedAlgorithmError(pki.oids[algorithmOid]);
    }
}

function getMessageDigest(attribute: asn1.Asn1): string {
    const parent = getParent(attribute, pki.oids["messageDigest"]);
    if (!parent) {
        throw new InvalidPkcs7EnvelopeError();
    }
    return <string>(<asn1.Asn1>(<asn1.Asn1>parent.value[1]).value[0]).value;
}

function getParent(attribute: asn1.Asn1, searchedOid: string): asn1.Asn1 | undefined {
    if (isOid(attribute, searchedOid)) {
        return attribute; // no parent so return myself
    } else if (Array.isArray(attribute.value)) {
        let index = 0;
        do {
            if (getParent(attribute.value[index], searchedOid)) {
                return attribute.value[index];
            }
            index++;
        } while (index < attribute.value.length);
    }
}

function isOid(attr: asn1.Asn1, oid: string): boolean {
    return attr.type === asn1.Type.OID && asn1.derToOid(new util.ByteStringBuffer(<string>attr.value)) === oid;
}
