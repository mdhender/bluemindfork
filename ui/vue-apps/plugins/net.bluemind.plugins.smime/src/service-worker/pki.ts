import { pki } from "node-forge";
import { PKIStatus } from "../lib/constants";
import { ExpiredCertificateError, InvalidKeyError, InvalidCertificateError } from "./exceptions";

import db from "./SMimeDB";

let PRIVATE_KEY: pki.rsa.PrivateKey;
let CERTIFICATE: pki.Certificate;

export async function getMyPrivateKey(): Promise<pki.rsa.PrivateKey> {
    if (!PRIVATE_KEY) {
        await load();
    }
    return PRIVATE_KEY;
}
export async function getMyCertificate(): Promise<pki.Certificate> {
    if (!CERTIFICATE) {
        await load();
    }
    return CERTIFICATE;
}

export async function getCertificate(email: string): Promise<pki.Certificate | undefined> {
    // TODO
    return Promise.resolve(undefined);
}

export function checkCertificateValidity(certificate: pki.Certificate, sendingDate: Date) {
    // TODO ? check if certificate is not revoked
    // TODO ? check chain of certificate (see pki.verifyCertificateChain)
    const isExpired = certificate.validity.notBefore > sendingDate || certificate.validity.notAfter < sendingDate;
    if (isExpired) {
        throw new ExpiredCertificateError();
    }
}

async function load(): Promise<void> {
    const pkiStatus = await db.getPKIStatus();
    if ((pkiStatus & PKIStatus.OK) === PKIStatus.OK) {
        try {
            const key = await ((await db.getPrivateKey()) as Blob).text();
            PRIVATE_KEY = pki.privateKeyFromPem(key);
        } catch (error) {
            throw new InvalidKeyError(error);
        }
        try {
            CERTIFICATE = pki.certificateFromPem(await ((await db.getCertificate()) as Blob).text());
        } catch (error) {
            throw new InvalidCertificateError(error);
        }
    } else if (pkiStatus & PKIStatus.PRIVATE_KEY_OK) {
        throw new InvalidCertificateError();
    } else {
        throw new InvalidKeyError();
    }
}
