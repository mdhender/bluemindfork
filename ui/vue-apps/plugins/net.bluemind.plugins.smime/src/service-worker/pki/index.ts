import { pki } from "node-forge";
import { AddressBooksClient, AddressBookClient, VCardInfo } from "@bluemind/addressbook.api";
import { searchVCardsHelper } from "@bluemind/contact";
import { ItemContainerValue } from "@bluemind/core.container.api";
import {
    checkBasicConstraints,
    checkExtendedKeyUsage,
    checkRecipientEmail,
    checkRevoked,
    checkSmimeUsage,
    getCaCerts
} from "./cert";
import { PKIStatus } from "../../lib/constants";
import session from "../environnment/session";
import {
    CertificateRecipientNotFoundError,
    InvalidCertificateError,
    InvalidKeyError,
    KeyNotFoundError,
    MyCertificateNotFoundError,
    UntrustedCertificateError
} from "../../lib/exceptions";
import { CheckOptions } from "../../types";
import db from "./SMimePkiDB";

export async function getCertificate(email: string): Promise<pki.Certificate> {
    const sid = await session.sid;
    const contactsInfo = await new AddressBooksClient(sid).search(searchVCardsHelper(email));
    if (contactsInfo.total === 0) {
        throw new CertificateRecipientNotFoundError(email);
    }
    const byContainers = groupByContainer(contactsInfo.values);

    let pem: string | undefined;
    for (const [containerUid, uids] of Object.entries(byContainers)) {
        const contacts = await new AddressBookClient(sid, containerUid).multipleGet(uids);
        const contact = contacts.find(contact => contact.value.security?.key?.value);
        pem = contact?.value?.security?.key?.value;
        if (pem) {
            break;
        }
    }
    if (!pem) {
        throw new CertificateRecipientNotFoundError(email);
    }
    try {
        return pki.certificateFromPem(pem);
    } catch (error) {
        throw new InvalidCertificateError(error);
    }
}

export async function checkCertificate(certificate: pki.Certificate, options?: CheckOptions) {
    try {
        const caCerts = await getCaCerts();
        if (caCerts.length === 0) {
            throw "could not find any trusted CA certificates";
        }
        const caStore = pki.createCaStore(caCerts.map(item => item.value.cert));
        pki.verifyCertificateChain(caStore, [certificate], { validityCheckDate: options?.date || new Date() });
        checkBasicConstraints(certificate);
        checkExtendedKeyUsage(certificate);
        if (options?.expectedAddress) {
            checkRecipientEmail(certificate, options.expectedAddress);
        }
        if (options?.smimeUsage) {
            checkSmimeUsage(certificate, options.smimeUsage);
        }
        await checkRevoked(certificate, options?.date);
    } catch (error: unknown) {
        if (typeof error === "string") {
            throw new UntrustedCertificateError(error);
        } else if ((<pki.ForgePkiCertificateError>error).error?.startsWith("forge.pki.")) {
            throw new UntrustedCertificateError((<pki.ForgePkiCertificateError>error).error);
        }
        throw error;
    }
}

interface Cache {
    CERTIFICATE: pki.Certificate | null;
    PRIVATE_KEY: pki.rsa.PrivateKey | null;
}

const cache: Cache = {
    CERTIFICATE: null,
    PRIVATE_KEY: null
};
export async function getMyStatus() {
    return await db.getPKIStatus();
}
export async function clear() {
    await db.clearMyCertAndKey();
    cache.CERTIFICATE = null;
    cache.PRIVATE_KEY = null;
    await db.clearRevocations();
}
export async function getMyPrivateKey(): Promise<pki.rsa.PrivateKey> {
    if (!cache.PRIVATE_KEY) {
        const pkiStatus = await db.getPKIStatus();
        if (pkiStatus & PKIStatus.OK || pkiStatus & PKIStatus.PRIVATE_KEY_OK) {
            try {
                const key = await ((await db.getPrivateKey()) as Blob).text();
                cache.PRIVATE_KEY = pki.privateKeyFromPem(key);
            } catch (error) {
                throw new InvalidKeyError(error);
            }
        } else {
            throw new KeyNotFoundError();
        }
    }
    return <pki.rsa.PrivateKey>cache.PRIVATE_KEY;
}

export async function setMyPrivateKey(blob: Blob) {
    await db.setPrivateKey(blob);
    cache.PRIVATE_KEY = pki.privateKeyFromPem(await blob.text());
}

export async function getMyCertificate() {
    if (!cache.CERTIFICATE) {
        const pkiStatus = await db.getPKIStatus();
        if (pkiStatus & PKIStatus.OK || pkiStatus & PKIStatus.CERTIFICATE_OK) {
            try {
                const cert = await (<Blob>await db.getCertificate()).text();
                cache.CERTIFICATE = pki.certificateFromPem(cert);
            } catch (error) {
                throw new InvalidCertificateError(error);
            }
        } else {
            throw new MyCertificateNotFoundError();
        }
    }
    return <pki.Certificate>cache.CERTIFICATE;
}

export async function setMyCertificate(blob: Blob) {
    await db.setCertificate(blob);
    cache.CERTIFICATE = pki.certificateFromPem(await blob.text());
}

function groupByContainer(contactsInfo: ItemContainerValue<VCardInfo>[]) {
    const byContainers: { [container: string]: string[] } = contactsInfo.reduce(
        (byContainer: { [containerUid: string]: string[] }, contactInfo: ItemContainerValue<VCardInfo>) => {
            if (!byContainer[contactInfo.containerUid]) {
                byContainer[contactInfo.containerUid] = [];
            }
            byContainer[contactInfo.containerUid].push(contactInfo.uid!);
            return byContainer;
        },
        {}
    );
    return byContainers;
}
