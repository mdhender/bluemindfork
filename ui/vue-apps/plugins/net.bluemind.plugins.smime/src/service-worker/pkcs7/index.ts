import { pkcs7, pki, asn1, util } from "node-forge";
import { binaryToArrayBuffer } from "@bluemind/arraybuffer";
import {
    InvalidCertificateError,
    DecryptError,
    EncryptError,
    SignError,
    UnmatchedCertificateError
} from "../exceptions";
import { checkMessageIntegrity, checkSignatureValidity, getSignedDataEnvelope, getSigningTime } from "./verify";
import { checkCertificateValidity } from "../pki/";

export async function decrypt(
    data: Blob,
    privateKey: pki.rsa.PrivateKey,
    certificate: pki.Certificate
): Promise<string | undefined> {
    const buffer = await data.arrayBuffer();
    const text = asn1.fromDer(new util.ByteStringBuffer(buffer));
    const envelope = <pkcs7.Captured<pkcs7.PkcsEnvelopedData>>pkcs7.messageFromAsn1(text);
    let recipient;
    try {
        recipient = envelope.findRecipient(certificate);
    } catch (error) {
        throw new InvalidCertificateError(error);
    }
    if (recipient) {
        try {
            envelope.decrypt(recipient, privateKey);
            return envelope.content?.toString();
        } catch (error) {
            throw new DecryptError(error);
        }
    } else {
        throw new UnmatchedCertificateError();
    }
}

export function encrypt(content: string, certificates: pki.Certificate[]): Blob {
    try {
        const envelope = pkcs7.createEnvelopedData();
        certificates.forEach(certificate => envelope.addRecipient(certificate));
        envelope.content = util.createBuffer(content);
        envelope.encrypt();
        const bytes = asn1.toDer(envelope.toAsn1()).getBytes();
        const buffer = binaryToArrayBuffer(bytes);
        return new Blob([buffer]);
    } catch (error) {
        throw new EncryptError(error);
    }
}

export async function verify(pkcs7: ArrayBuffer, toDigest: string) {
    const envelope = getSignedDataEnvelope(pkcs7);
    const signingTime = getSigningTime(envelope);
    const certificate = envelope.certificates[0];
    checkCertificateValidity(certificate, signingTime);
    checkSignatureValidity(envelope, certificate);
    checkMessageIntegrity(envelope, toDigest);
}

// create PKCS#7 signed data with authenticatedAttributes
// attributes include: PKCS#9 content-type, message-digest, and signing-time
export async function sign(content: string, myPrivateKey: pki.rsa.PrivateKey, myCert: pki.Certificate) {
    try {
        const p7 = pkcs7.createSignedData();
        p7.content = util.createBuffer(content);

        // FIXME: add CA cert ?
        p7.addCertificate(myCert);
        p7.addSigner({
            key: myPrivateKey,
            certificate: myCert,
            digestAlgorithm: pki.oids.sha256,
            authenticatedAttributes: [
                {
                    type: pki.oids.contentType,
                    value: pki.oids.data
                },
                {
                    type: pki.oids.messageDigest // value will be auto-populated at signing time
                },
                {
                    type: pki.oids.signingTime,
                    value: asn1.dateToUtcTime(new Date())
                }
            ]
        });

        p7.sign({ detached: true }); // PKCS#7 sign in detached mode: includes the signature and certificate without the signed data
        const pem = pkcs7.messageToPem(p7);
        return removePemLabel(pem);
    } catch (error) {
        throw new SignError(error);
    }
}

function removePemLabel(pem: string): string {
    const firstNewLine = pem.indexOf("\r\n");
    return pem.substring(firstNewLine + "\r\n".length, pem.length - "\r\n".length * 2 - "-----END PKCS7-----".length);
}

export default { decrypt, encrypt, verify, sign };
