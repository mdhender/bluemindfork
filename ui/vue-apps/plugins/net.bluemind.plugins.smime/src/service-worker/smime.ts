import { MailboxItemsClient, MailboxItem, MessageBody } from "@bluemind/backend.mail.api";
import { ItemValue } from "@bluemind/core.container.api";
import { MimeParser, MimeBuilder } from "@bluemind/mime";
import {
    CRYPTO_HEADERS,
    ENCRYPTED_HEADER_NAME,
    MULTIPART_SIGNED_MIME,
    PKCS7_MIMES,
    SIGNED_HEADER_NAME
} from "../lib/constants";
import session from "./environnment/session";
import { EncryptError, RecipientNotFoundError, SmimeErrors } from "./exceptions";
import extractSignedData from "./signedDataParser";
import pkcs7 from "./pkcs7/";
import { checkCertificateValidity, getMyCertificate, getMyPrivateKey } from "./pki/";
import { logger } from "./environnment/logger";

export function isEncrypted(item: ItemValue<MailboxItem>): boolean {
    return PKCS7_MIMES.includes(item.value!.body!.structure!.mime!);
}
export function isSigned(item: ItemValue<MailboxItem>): boolean {
    return item.value!.body!.structure!.mime === MULTIPART_SIGNED_MIME;
}

export async function decrypt(folderUid: string, item: ItemValue<MailboxItem>): Promise<ItemValue<MailboxItem>> {
    //TODO: Add a cache based on body guid
    try {
        const sid = await session.sid;
        const client = new MailboxItemsClient(sid, folderUid);
        const part = item.value!.body!.structure!;
        const key = await getMyPrivateKey();
        const certificate = await getMyCertificate();
        // FIXME: use correct date instead of internalDate
        checkCertificateValidity(certificate, new Date(item.value!.body!.date!));
        const data = await client.fetch(item.value!.imapUid!, part.address!, part.encoding!, part.mime!);
        const content = await pkcs7.decrypt(data, key, certificate);
        if (content) {
            const parser = await new MimeParser(part.address).parse(content);
            const parts = parser.getParts();
            for (const p of parts) {
                const content = parser.getPartContent(p.address!);
                savePart(folderUid, item.value!.imapUid!, p, content);
            }
            item.value!.body!.structure = parser.structure as MessageBody.Part;
        }
        setHeader(item, ENCRYPTED_HEADER_NAME, CRYPTO_HEADERS.DECRYPTED);
    } catch (error: unknown) {
        const errorCode = error instanceof SmimeErrors ? error.code : CRYPTO_HEADERS.UNKNOWN;
        setHeader(item, ENCRYPTED_HEADER_NAME, errorCode);
    }
    return item;
}

export async function encrypt(item: MailboxItem, folderUid: string): Promise<MailboxItem> {
    const encryptedItem = { ...item };
    const sid = await session.sid;
    const client = new MailboxItemsClient(sid, folderUid);
    let mimeTree: string | undefined;

    // TODO: get the certificates for all recipients
    try {
        const certificate = await getMyCertificate();
        if (encryptedItem.body.structure && encryptedItem.imapUid) {
            const getContentFn = async (p: MessageBody.Part): Promise<Uint8Array | null> => {
                if (!encryptedItem.imapUid || !p.address) {
                    return null;
                }
                const data: Blob = await client.fetch(encryptedItem.imapUid, p.address, p.encoding, p.mime);
                return new Uint8Array(await data.arrayBuffer());
            };

            try {
                mimeTree = await new MimeBuilder(getContentFn).build(encryptedItem.body.structure);
            } catch (error) {
                throw new EncryptError(error);
            }
            if (mimeTree) {
                const encryptedPart = pkcs7.encrypt(mimeTree, [certificate]);
                const address = await client.uploadPart(encryptedPart);
                const part = {
                    address,
                    charset: "utf-8",
                    encoding: "base64",
                    mime: PKCS7_MIMES[0]
                };
                encryptedItem.body.structure = part;
            }
        }
    } catch (error) {
        const errorCode = error instanceof SmimeErrors ? error.code : CRYPTO_HEADERS.UNKNOWN;
        setHeader({ value: encryptedItem }, ENCRYPTED_HEADER_NAME, errorCode);
    }
    return encryptedItem;
}

export async function verify(folderUid: string, item: ItemValue<MailboxItem>): Promise<ItemValue<MailboxItem>> {
    try {
        const client = new MailboxItemsClient(await session.sid, folderUid);
        const eml = await client.fetchComplete(item.value.imapUid!).then(eml => eml.text());
        const { toDigest, pkcs7Part } = extractSignedData(eml);
        await pkcs7.verify(pkcs7Part, toDigest, getSenderAddress(item));
        setHeader(item, SIGNED_HEADER_NAME, CRYPTO_HEADERS.VERIFIED);
    } catch (error) {
        logger.error(error);
        const errorCode = error instanceof SmimeErrors ? error.code : CRYPTO_HEADERS.UNKNOWN;
        setHeader(item, SIGNED_HEADER_NAME, errorCode);
    }
    return item;
}

export function sign() {
    return null;
}

function getSenderAddress(item: ItemValue<MailboxItem>): string {
    if (item.value.body.recipients) {
        const from = item.value.body.recipients.find(
            recipient => recipient.kind === MessageBody.RecipientKind.Originator
        );
        if (!from) {
            throw new RecipientNotFoundError();
        }
        return from.address || "";
    }
    return "";
}

//FIXME: This should be imported from a third party package
async function savePart(uid: string, imap: number, part: MessageBody.Part, content: ArrayBuffer | undefined) {
    const cache = await caches.open("part-cache");
    const { address } = part;
    if (address) {
        const request = new Request(constructCacheUrl(uid, imap, address));
        cache.put(request.url, new Response(content));
    }
}

function setHeader(item: ItemValue<MailboxItem>, headerName: string, headerValue: number): void {
    if (item.value.body?.headers) {
        const index = item.value.body.headers.findIndex(({ name }) => name === headerName);

        if (index === -1) {
            item.value.body.headers.push({ name: headerName, values: [headerValue.toString()] });
        } else {
            const currentValues = item.value.body.headers[index].values || [];
            const newValue = parseInt(currentValues[0]) | headerValue;
            item.value.body.headers[index] = { name: headerName, values: [newValue.toString()] };
        }
    }
}

function constructCacheUrl(folderUid: string, imapUid: number, address: string) {
    return `/api/mail_items/${folderUid}/part/${imapUid}/${address}`;
}

export default {
    isEncrypted,
    decrypt,
    encrypt,
    verify,
    sign
};
