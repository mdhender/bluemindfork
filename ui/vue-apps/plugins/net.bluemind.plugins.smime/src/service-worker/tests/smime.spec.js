import { MimeType } from "@bluemind/email";
import smime from "../smime";
import pkcs7 from "../pkcs7";
import pki from "../pki/";
import forge from "node-forge";
import {
    EncryptError,
    ExpiredCertificateError,
    InvalidSignatureError,
    RevokedCertificateError,
    SignError,
    UntrustedCertificateError,
    UnmatchedCertificateError
} from "../exceptions";
import {
    CRYPTO_HEADERS,
    ENCRYPTED_HEADER_NAME,
    SIGNED_HEADER_NAME,
    SMIME_ENCRYPTION_ERROR_PREFIX,
    SMIME_SIGNATURE_ERROR_PREFIX
} from "../../lib/constants";
import { getHeaderValue, isVerified } from "../../lib/helper";
import { readFile } from "./helpers";

jest.mock("../pki/", () => jest.fn);
jest.mock("@bluemind/mime", () => {
    return {
        ...jest.requireActual("@bluemind/mime"),
        MimeBuilder: () => ({
            build: () => "dummy structure"
        })
    };
});

const mockUploadPart = jest.fn(() => Promise.resolve("address"));
jest.mock("@bluemind/backend.mail.api", () => ({
    ...jest.requireActual("@bluemind/backend.mail.api"),
    MailboxItemsClient: () => ({
        fetch: () => Promise.resolve("data"),
        uploadPart: mockUploadPart
    })
}));
jest.mock("../pkcs7", () => jest.fn);

let mockCache = {};
global.caches = {
    open: () => {
        return Promise.resolve({
            put: (request, response) => {
                mockCache[request.url] = response;
            }
        });
    }
};
jest.mock("../environnment/session", () =>
    Promise.resolve({
        json: () =>
            Promise.resolve({
                login: "mathilde.michau@blue-mind.net",
                sid: "58a1ee1b-0c30-492c-a83f-4396f0a24730"
            })
    })
);

class MockRequest {
    constructor(url) {
        this.url = url;
    }
}
class MockResponse {}
global.Request = MockRequest;
global.Response = MockResponse;

const mainEncrypted = {
    value: {
        body: {
            structure: {
                address: "1",
                mime: "application/pkcs7-mime"
            },
            headers: [],
            date: 1668534530000
        },
        imapUid: 99
    }
};

const unencrypted = {
    value: {
        body: {
            structure: {
                mime: "multipart/alternative",
                address: "TEXT",
                children: [{ mime: "text/plain" }, { mime: "text/html" }]
            },
            headers: [],
            date: 1668534530
        }
    }
};

const certificateTxt = readTxt("documents/certificate");
const mockCertificate = forge.pki.certificateFromPem(certificateTxt);

describe("smime", () => {
    let item;
    beforeEach(() => {
        item = {
            body: {
                date: 1671032461777,
                subject: "Mail",
                headers: [],
                recipients: [
                    {
                        kind: "Primary",
                        dn: "math",
                        address: "math@devenv.blue"
                    },
                    {
                        kind: "Originator",
                        dn: "math",
                        address: "math@devenv.blue"
                    }
                ],
                messageId: "<lbntihyw.j2pop9bobhc0@devenv.blue>",
                structure: {
                    mime: "multipart/alternative",
                    children: [
                        {
                            mime: "text/plain",
                            address: "06a5ccf7-4094-4c7f-8533-eb99b072b28d",
                            encoding: "quoted-printable",
                            charset: "utf-8"
                        },
                        {
                            mime: "text/html",
                            address: "9ba724be-a4b9-4679-be0c-1c5207401d38",
                            encoding: "quoted-printable",
                            charset: "utf-8"
                        }
                    ]
                }
            },
            imapUid: 424,
            flags: ["\\Seen"]
        };

        pkcs7.decrypt = () => Promise.resolve("content");
        pki.getMyPrivateKey = () => Promise.resolve("PrivateKey");
        pki.getMyCertificate = () =>
            Promise.resolve({
                validity: {
                    notBefore: new Date("2022-09-25T13:43:26.000Z"),
                    notAfter: new Date("2023-09-25T13:43:26.000Z")
                }
            });
        pki.getCertificate = () => [];
        pki.checkCertificateValidity = () => {};
        jest.clearAllMocks();
    });

    describe("isEncrypted", () => {
        test("return true if the message main part is crypted", () => {
            const isEncrypted = smime.isEncrypted(mainEncrypted.value.body.structure);
            expect(isEncrypted).toBe(true);
        });
        test("return true if a subpart of the message is crypted", () => {});
        test("return true if multiple subpart of the message is crypted", () => {});
        test("return false if there is not crypted subpart", () => {
            const isEncrypted = smime.isEncrypted(unencrypted);
            expect(isEncrypted).toBe(false);
        });
    });
    describe("decrypt", () => {
        beforeEach(() => {
            mainEncrypted.value.body.headers = [];
            mockCache = {};
        });
        test("adapt message body structure when the main part is encrypted", async () => {
            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(item.value.body.structure).toEqual(expect.objectContaining({ mime: "text/plain", address: "1" }));
        });
        test("add decrypted parts content to part cache", async () => {
            const mockEmlMultipleParts = readEml("eml/unencrypted");
            pkcs7.decrypt = () => Promise.resolve(mockEmlMultipleParts);

            await smime.decrypt("uid", mainEncrypted);
            expect(mockCache).toMatchSnapshot();
        });
        test("raise an error if the certificate is expired", async () => {
            pki.checkCertificateValidity = () => {
                throw new ExpiredCertificateError();
            };
            const old = {
                value: {
                    body: {
                        structure: {
                            address: "1",
                            mime: "application/pkcs7-mime"
                        },
                        headers: [],
                        date: 1000
                    },
                    imapUid: 99
                }
            };
            const { item } = await smime.decrypt("uid", old);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.EXPIRED_CERTIFICATE).toBeTruthy();
        });
        test("add a header if the message is crypted", async () => {
            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item)).toBeTruthy();
        });
        test("add a header if the message is correcty decrypted", async () => {
            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.OK).toBeTruthy();
        });
        test("add a header if the message cannot be decrypted because private key or certificate are expired", async () => {
            pkcs7.decrypt = () => Promise.reject(new ExpiredCertificateError());

            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.EXPIRED_CERTIFICATE).toBeTruthy();
        });
        test("add a header if the message cannot be decrypted because private key or certificate are revoked", async () => {
            pkcs7.decrypt = () => Promise.reject(new RevokedCertificateError());

            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.REVOKED_CERTIFICATE).toBeTruthy();
        });
        test("add a header if the message cannot be decrypted because private key or certificate are not trusted", async () => {
            pkcs7.decrypt = () => Promise.reject(new UntrustedCertificateError());

            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.UNTRUSTED_CERTIFICATE).toBeTruthy();
        });
        test("add a header if the given certificate does not match any recipient", async () => {
            pkcs7.decrypt = () => Promise.reject(new UnmatchedCertificateError());

            const { item } = await smime.decrypt("uid", mainEncrypted);
            expect(getCryptoHeaderCode(item) & CRYPTO_HEADERS.UNMATCHED_CERTIFICATE).toBeTruthy();
        });
    });
    describe("encrypt", () => {
        beforeEach(() => {
            pkcs7.encrypt = () => "encrypted";
            pki.getMyCertificate = () => Promise.resolve(mockCertificate);
        });
        test("adapt message body structure and upload new encrypted part when the main part has to be encrypted ", async () => {
            const structure = await smime.encrypt(item, "folderUid");
            expect(mockUploadPart).toHaveBeenCalled();
            expect(structure.body.structure.address).toBe("address");
            expect(structure.body.structure.mime).toBe(MimeType.PKCS_7);
        });
        test("raise an error if the message cannot be encrypted", async () => {
            pkcs7.encrypt = () => {
                throw new EncryptError();
            };
            try {
                await smime.encrypt(item, "folderUid");
            } catch (error) {
                expect(error).toContain(SMIME_ENCRYPTION_ERROR_PREFIX);
            }
        });
    });
    describe("sign", () => {
        beforeEach(() => {
            pkcs7.sign = jest.fn(() => Promise.resolve("b64"));
        });
        test("adapt body structure and upload full eml", async () => {
            const signedItem = await smime.sign(item, "folderUid");
            const multipartSigned = signedItem.body.structure;
            expect(multipartSigned.mime).toBe("message/rfc822");
            expect(multipartSigned.children.length).toBe(0);
            expect(mockUploadPart).toHaveBeenCalledTimes(1);
        });
        test("raise an error if the message cannot be signed", async done => {
            try {
                pkcs7.sign = () => {
                    throw new SignError();
                };
                await smime.sign(item, "folderUid");
                done.fail();
            } catch (error) {
                expect(error).toContain(SMIME_SIGNATURE_ERROR_PREFIX);
                expect(error).toContain(CRYPTO_HEADERS.SIGN_FAILURE);
                done();
            }
        });
    });
    describe("verify", () => {
        const getEml = () => readEml("eml/signed_only/valid");
        beforeEach(() => {
            pkcs7.verify = () => Promise.resolve();
        });
        test("add a OK header if item is successfuly verified", async () => {
            const itemValue = { value: item };
            const verified = await smime.verify("folderUid", itemValue, getEml);
            expect(isVerified(verified.value.body.headers)).toBe(true);
        });
        test("add a KO header if item cant be verified", async () => {
            pkcs7.verify = () => {
                throw new InvalidSignatureError();
            };
            const itemValue = { value: item };
            const verified = await smime.verify("folderUid", itemValue, getEml);
            expect(isVerified(verified.value.body.headers)).toBe(false);
            const headerValue = getHeaderValue(item.body.headers, SIGNED_HEADER_NAME);
            expect(Boolean(headerValue & CRYPTO_HEADERS.INVALID_SIGNATURE)).toBe(true);
        });
    });
});

function readEml(file) {
    return readFile(`${file}.eml`);
}

function getCryptoHeaderCode(item) {
    const code = item.value.body.headers.find(h => h.name === ENCRYPTED_HEADER_NAME).values[0];
    return parseInt(code);
}

function readTxt(file) {
    return readFile(`${file}.txt`);
}
