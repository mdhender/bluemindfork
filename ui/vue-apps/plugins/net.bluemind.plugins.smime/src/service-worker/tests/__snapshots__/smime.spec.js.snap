// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smime decrypt add decrypted parts content to part cache 1`] = `
Object {
  "99/1.1.1": MockResponse {},
  "99/1.1.2.1": MockResponse {},
  "99/1.1.2.2": MockResponse {},
  "99/1.2": MockResponse {},
}
`;
